File: app/src/main/java/com/quibbler/sevenmusic/utils/ColorFilterTransformation.java
package com.quibbler.sevenmusic.utils;

import android.graphics.Bitmap;
import android.graphics.Canvas;
import android.graphics.Paint;
import android.graphics.PorterDuff;
import android.graphics.PorterDuffColorFilter;

import androidx.annotation.NonNull;

import com.bumptech.glide.load.engine.bitmap_recycle.BitmapPool;
import com.bumptech.glide.load.resource.bitmap.BitmapTransformation;

import java.security.MessageDigest;

/**
  *
  * Package:        com.quibbler.sevenmusic.utils
  * ClassName:      ColorFilterTransformation
  * Description:    加载图片颜色变换，配合glide使用
  * Author:         lishijun
  * CreateDate:     2019/10/10 10:36
 */
public class ColorFilterTransformation extends BitmapTransformation {

    private static final int VERSION = 1;
    private static final String ID = "ColorFilterTransformation." + VERSION;

    private int mColor;

    public ColorFilterTransformation(int color) {
        mColor = color;
    }

    @Override
    protected Bitmap transform(@NonNull BitmapPool pool, @NonNull Bitmap toTransform, int outWidth, int outHeight) {
        int width = toTransform.getWidth();
        int height = toTransform.getHeight();

        Bitmap.Config config =
                toTransform.getConfig() != null ? toTransform.getConfig() : Bitmap.Config.ARGB_8888;
        Bitmap bitmap = pool.get(width, height, config);
        if (bitmap == null) {
            bitmap = Bitmap.createBitmap(width, height, config);
        }
        Canvas canvas = new Canvas(bitmap);
        Paint paint = new Paint();
        paint.setAntiAlias(true);
        paint.setColorFilter(new PorterDuffColorFilter(mColor, PorterDuff.Mode.SRC_ATOP));
        canvas.drawBitmap(toTransform, 0, 0, paint);

        return bitmap;
    }

    @Override public void updateDiskCacheKey(@NonNull MessageDigest messageDigest) {
        messageDigest.update((ID + mColor).getBytes(CHARSET));
    }
}

File: app/src/main/java/com/quibbler/sevenmusic/utils/ColorFilterTransformation.kt
package com.quibbler.sevenmusic.utils

import android.graphics.Bitmap
import android.graphics.Canvas
import android.graphics.Paint
import android.graphics.PorterDuff
import android.graphics.PorterDuffColorFilter
import com.bumptech.glide.load.engine.bitmap_recycle.BitmapPool
import com.bumptech.glide.load.resource.bitmap.BitmapTransformation
import java.security.MessageDigest

/**
 * 
 * Package:        com.quibbler.sevenmusic.utils
 * ClassName:      ColorFilterTransformation
 * Description:    加载图片颜色变换，配合glide使用
 * Author:         lishijun
 * CreateDate:     2019/10/10 10:36
 */
class ColorFilterTransformation(color: Int) : BitmapTransformation() {
    private val mColor: Int

    init {
        mColor = color
    }

    override fun transform(
        pool: BitmapPool,
        toTransform: Bitmap,
        outWidth: Int,
        outHeight: Int
    ): Bitmap {
        val width = toTransform.getWidth()
        val height = toTransform.getHeight()

        val config =
            toTransform.getConfig() ?: Bitmap.Config.ARGB_8888
        var bitmap = pool.get(width, height, config)
        if (bitmap == null) {
            bitmap = Bitmap.createBitmap(width, height, config)
        }
        val canvas = Canvas(bitmap)
        val paint = Paint()
        paint.setAntiAlias(true)
        paint.setColorFilter(PorterDuffColorFilter(mColor, PorterDuff.Mode.SRC_ATOP))
        canvas.drawBitmap(toTransform, 0f, 0f, paint)

        return bitmap
    }

    override fun updateDiskCacheKey(messageDigest: MessageDigest) {
        messageDigest.update((ID + mColor).toByteArray(CHARSET))
    }

    companion object {
        private const val VERSION = 1
        private val ID = "ColorFilterTransformation." + VERSION
    }
}
